cozy/synthesis/core.py

```diff
@@ -210,6 +214,9 @@ def improve(
     blacklist = {}
 
     while True:
+        if improvement_limit.value != -1 and improve_count.value >= improvement_limit.value:
```

Comment:
This might have an off-by-one error and shows how racy Cozy could be. If I set limit to `10`, the reported run is actually `13`.

```python
         if improve_count is not None:
            with improve_count.get_lock():
                if improvement_limit.value == -1 or improve_count.value < improvement_limit.value:
                    improve_count.value += 1
                else:
                    return
```

Reply:
note that the above snippet is in the beginning of the loop, meaning "check and consume the quota before actually use it".